### ISSUE TITLE
Use correct integer types in Cython

### ISSUE THREAD

**Issue by niboshi** [cupy/cupy#1913]
Currently incorrect integer types are used in Cython implementation.
It's less portable and perhaps more importantly, confusing for developers.

For example:
* Pointers should be `intptr_t`.
* Memory sizes should be `size_t`.
* Memory offsets should be `ptrdiff_t`.

**Comment by leofang** [comment]
Hi @ninoshi, when I was working on PRs that would touch the driver wrappers, I noticed that in `driver.pxd`/`driver.pyx` there are many types that should have been changed from `size_t` to `intptr_t` but were left untouched in #1952, which made me thought that using `size_t` was the deliberately chosen convention for drivers, but looking at this issue I doubt it's the case. 

Specifically, I am referring to the driver functions involving the following types:
https://github.com/cupy/cupy/blob/c670b6f6f1282665c9985180c10f498d8591f909/cupy/cuda/driver.pxd#L12-L18
all of which are just pointers but were represented using `size_t`. Is this an oversight? If so I can make a PR to fix them.
